src/graph/executor/algo/ShortestPathBase.h

```diff
@@ -0,0 +1,98 @@
+// Copyright (c) 2022 vesoft inc. All rights reserved.
+//
+// This source code is licensed under Apache 2.0 License.
+#ifndef GRAPH_EXECUTOR_ALGO_SHORTESTPATHBASE_H_
+#define GRAPH_EXECUTOR_ALGO_SHORTESTPATHBASE_H_
+
+#include "graph/planner/plan/Algo.h"
+
+using nebula::storage::StorageRpcResponse;
+using nebula::storage::cpp2::GetNeighborsResponse;
+using RpcResponse = StorageRpcResponse<GetNeighborsResponse>;
+using PropRpcResponse = StorageRpcResponse<nebula::storage::cpp2::GetPropResponse>;
+namespace nebula {
+namespace graph {
+class ShortestPathBase {
+ public:
+  ShortestPathBase(const ShortestPath* node,
+                   QueryContext* qctx,
+                   std::unordered_map<std::string, std::string>* stats)
+      : pathNode_(node), qctx_(qctx), stats_(stats) {
+    singleShortest_ = node->singleShortest();
+    maxStep_ = node->stepRange()->max();
+  }
+
+  virtual ~ShortestPathBase() {}
+
+  virtual folly::Future<Status> execute(const std::unordered_set<Value>& startVids,
+                                        const std::unordered_set<Value>& endVids,
+                                        DataSet* result) = 0;
+
+  using DstVid = Value;
+  // start vid of the path
+  using StartVid = Value;
+  // end vid of the path
+  using EndVid = Value;
+  // save the starting vertex and the corresponding edge. eg [vertex(a), edge(a->b)]
+  using CustomStep = Row;
+
+ protected:
+  folly::Future<std::vector<Value>> getMeetVidsProps(const std::vector<Value>& meetVids);
+
+  std::vector<Value> handlePropResp(PropRpcResponse&& resps);
+
+  Status handleErrorCode(nebula::cpp2::ErrorCode code, PartitionID partId) const;
+
+  void addStats(RpcResponse& resp,
+                std::unordered_map<std::string, std::string>* stats,
+                size_t stepNum,
+                int64_t timeInUSec,
+                bool reverse) const;
+
+  void addStats(PropRpcResponse& resp,
+                std::unordered_map<std::string, std::string>* stats,
+                int64_t timeInUSec) const;
+
+  template <typename Resp>
+  StatusOr<Result::State> handleCompleteness(const storage::StorageRpcResponse<Resp>& rpcResp,
```

Comment:
Inherit StorageAccessExecutor and reuse these functions.

Reply:
If inheriting from storageAccessExector, batchShortestPath and singleShortestPath are different operators, which is not what we expect

Reply:
Why different?

Reply:
they are two different operators. If they both inherit storageAccessExector,